// next/dist/diagnostics/build-diagnostics.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
0 && (module.exports = {
    recordFrameworkVersion: null,
    updateBuildDiagnostics: null
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    recordFrameworkVersion: function() {
        return recordFrameworkVersion;
    },
    updateBuildDiagnostics: function() {
        return updateBuildDiagnostics;
    }
});
const _promises = require("fs/promises");
const _path = require("path");
const _shared = require("../trace/shared");
const DIAGNOSTICS_DIR = 'diagnostics';
const DIAGNOSTICS_FILE = 'build-diagnostics.json';
async function getDiagnosticsDir() {
    const distDir = _shared.traceGlobals.get('distDir');
    const diagnosticsDir = (0, _path.join)(distDir, DIAGNOSTICS_DIR);
    await (0, _promises.mkdir)(diagnosticsDir, {
        recursive: true
    });
    return diagnosticsDir;
}
async function recordFrameworkVersion(version) {
    const diagnosticsDir = await getDiagnosticsDir();
    const frameworkVersionFile = (0, _path.join)(diagnosticsDir, 'framework.json');
    await (0, _promises.writeFile)(frameworkVersionFile, JSON.stringify({
        name: 'Next.js',
        version
    }));
}
async function updateBuildDiagnostics(diagnostics) {
    const diagnosticsDir = await getDiagnosticsDir();
    const diagnosticsFile = (0, _path.join)(diagnosticsDir, DIAGNOSTICS_FILE);
    const existingDiagnostics = JSON.parse(await (0, _promises.readFile)(diagnosticsFile, 'utf8').catch(()=>'{}'));
    const updatedBuildOptions = {
        ...existingDiagnostics.buildOptions ?? {},
        ...diagnostics.buildOptions ?? {}
    };
    const updatedDiagnostics = {
        ...existingDiagnostics,
        ...diagnostics,
        buildOptions: updatedBuildOptions
    };
    await (0, _promises.writeFile)(diagnosticsFile, JSON.stringify(updatedDiagnostics, null, 2));
}
